src/reactive-handler.ts

```diff
@@ -1,74 +1,35 @@
-import {
-    toString,
-    isUndefined,
-    unwrap,
-    ArrayConcat,
-    isArray,
-    ObjectDefineProperty,
-    getPrototypeOf,
-    isExtensible,
-    getOwnPropertyDescriptor,
-    getOwnPropertyNames,
-    getOwnPropertySymbols,
-    preventExtensions,
-    hasOwnProperty,
-} from './shared';
+import { toString, isArray, unwrap, isExtensible, preventExtensions, ObjectDefineProperty, freeze } from './shared';
+import { BaseProxyHandler, ReactiveMembraneShadowTarget } from './base-handler';
 
-import {
-    ReactiveMembrane,
-    ReactiveMembraneShadowTarget,
-    wrapDescriptor,
-} from './reactive-membrane';
+const getterMap = new WeakMap<() => any, () => any>();
+const setterMap = new WeakMap<(v: any) => void, (v: any) => void>();
 
-function wrapValue(membrane: ReactiveMembrane, value: any): any {
-    return membrane.valueIsObservable(value) ? membrane.getProxy(value) : value;
-}
-
-/**
- * Unwrap property descriptors will set value on original descriptor
- * We only need to unwrap if value is specified
- * @param descriptor external descrpitor provided to define new property on original value
- */
-function unwrapDescriptor(descriptor: PropertyDescriptor): PropertyDescriptor {
-    if (hasOwnProperty.call(descriptor, 'value')) {
-        descriptor.value = unwrap(descriptor.value);
+export class ReactiveProxyHandler extends BaseProxyHandler {
+    wrapValue(value: any): any {
+        return this.membrane.getProxy(value);
     }
-    return descriptor;
-}
-
-function lockShadowTarget(membrane: ReactiveMembrane, shadowTarget: ReactiveMembraneShadowTarget, originalTarget: any): void {
-    const targetKeys = ArrayConcat.call(getOwnPropertyNames(originalTarget), getOwnPropertySymbols(originalTarget));
-    targetKeys.forEach((key: PropertyKey) => {
-        let descriptor = getOwnPropertyDescriptor(originalTarget, key) as PropertyDescriptor;
-
-        // We do not need to wrap the descriptor if configurable
-        // Because we can deal with wrapping it when user goes through
-        // Get own property descriptor. There is also a chance that this descriptor
-        // could change sometime in the future, so we can defer wrapping
-        // until we need to
-        if (!descriptor.configurable) {
-            descriptor = wrapDescriptor(membrane, descriptor, wrapValue);
+    wrapGetter(originalGet: () => any): () => any {
+        if (getterMap.has(originalGet)) {
+            return getterMap.get(originalGet) as () => any;
```

Comment:
This type issue exhibit should not be worked around since it might help with performance.

```suggestion
            const wrappedGetter = getterMap.get(originalGet);
            if (wrappedGetter !== undefined) {
                return wrappedGetter;
            }
```